File: android/app/src/main/java/com/vocahq/vocaphone/core/PairingPayload.kt
package com.vocahq.vocaphone.core

import org.json.JSONObject

/**
 * Shared phone-pairing document encoded in the gateway WebUI QR.
 *
 * Wire format (version 1):
 * `{"v":1,"url":"http://192.168.1.20:8765","token":"..."}`
 */
object PairingPayload {

    const val VERSION = 1

    data class Parsed(val url: String, val token: String)

    sealed interface Result {
        data class Ok(val parsed: Parsed) : Result
        data class Err(val reason: String) : Result
    }

    fun parse(raw: String): Result {
        val text = raw.trim()
        if (text.isEmpty()) return Result.Err("The pairing code is empty.")

        val data = try {
            JSONObject(text)
        } catch (_: Exception) {
            return Result.Err("This is not a VocaPhone pairing code.")
        }

        val version = when {
            data.has("v") -> data.optInt("v", -1)
            data.has("version") -> data.optInt("version", -1)
            else -> -1
        }
        if (version != VERSION) {
            return Result.Err("This pairing code uses an unsupported version.")
        }

        val url = data.optString("url", "").trim()
        val token = data.optString("token", "").trim()
        if (url.isEmpty()) return Result.Err("The pairing code is missing a gateway address.")
        if (token.isEmpty()) return Result.Err("The pairing code is missing a bearer token.")
        if (token.length < 32) {
            return Result.Err("The pairing token looks too short to be valid.")
        }

        return when (val validation = GatewayEndpoint.validate(url)) {
            is GatewayEndpoint.Validation.Invalid -> Result.Err(validation.reason)
            is GatewayEndpoint.Validation.Valid -> Result.Ok(Parsed(validation.url, token))
        }
    }

    /** Encode for tests and any host-side previews that share this format. */
    fun encode(url: String, token: String): String {
        val validation = GatewayEndpoint.validate(url)
        require(validation is GatewayEndpoint.Validation.Valid) {
            (validation as GatewayEndpoint.Validation.Invalid).reason
        }
        require(token.isNotBlank()) { "Token must not be empty." }
        return JSONObject()
            .put("v", VERSION)
            .put("url", validation.url)
            .put("token", token.trim())
            .toString()
    }
}

File: ios/VocaPhoneShared/PairingPayload.swift
import Foundation

/// The versioned document encoded in the gateway WebUI's phone-pairing QR.
///
/// Version 1 wire format:
/// `{"v":1,"url":"http://192.168.1.20:8765","token":"..."}`
enum PairingPayload {
    static let version = 1

    struct Value: Equatable {
        let url: URL
        let token: String
    }

    enum Result: Equatable {
        case success(Value)
        case failure(String)
    }

    static func parse(_ rawValue: String) -> Result {
        let rawValue = rawValue.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !rawValue.isEmpty else {
            return .failure("The pairing code is empty.")
        }

        let document: Document
        do {
            document = try JSONDecoder().decode(Document.self, from: Data(rawValue.utf8))
        } catch {
            return .failure("This is not a vocaphone pairing code.")
        }

        guard document.supportedVersion == version else {
            return .failure("This pairing code uses an unsupported version.")
        }

        let urlString = document.url.trimmingCharacters(in: .whitespacesAndNewlines)
        let token = document.token.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !urlString.isEmpty else {
            return .failure("The pairing code is missing a gateway address.")
        }
        guard !token.isEmpty else {
            return .failure("The pairing code is missing a bearer token.")
        }
        guard token.count >= 32 else {
            return .failure("The pairing token looks too short to be valid.")
        }
        guard let url = GatewayEndpoint.validatedURL(from: urlString) else {
            return .failure("The pairing code has an invalid gateway address.")
        }
        if GatewayEndpoint.usesUnencryptedHTTP(url),
           !GatewayEndpoint.isPrivateNetworkHost(url.host ?? "")
        {
            let host = url.host ?? "This host"
            return .failure(
                "\(host) is reachable from the public internet, so it needs https://."
            )
        }

        return .success(Value(url: url, token: token))
    }
}

private extension PairingPayload {
    struct Document: Decodable {
        let v: Int?
        let legacyVersion: Int?
        let url: String
        let token: String

        enum CodingKeys: String, CodingKey {
            case v
            case legacyVersion = "version"
            case url
            case token
        }

        var supportedVersion: Int? { v ?? legacyVersion }
    }
}
